### PR TITLE
Doc 1067 record notify errors in db

### PR DESCRIPTION
**Pull request by suityou01**
Changes to the way in which the notify client is called which include 

A fix to the retry function to make sure it uses async/await pattern when recursing.
A new thunkify method for turning function calls into thunks (for retrying).
A new promisify method for turning synchronous methods into asynchronous.
Some new columns in the failure table to record any error information coming from notify.
Calls to notify now retry 3 times before failing.

### CODE REVIEW

File: app/monitoring/create-failure.js [hunk 1]
```diff
@@ -1,10 +1,13 @@
 const db = require('../data')
 
-const createFailure = async (deliveryId, reason, transaction) => {
+const createFailure = async (deliveryId, errorObject, timestamp, transaction) => {
   await db.failure.create({
-    deliveryId,
-    reason,
-    failed: new Date()
+    deliveryId: deliveryId,
```
Comment (samplackett) [inline app/monitoring/create-failure.js:5]: believe this one can be reverted back to just deliveryId for the shorthand. I think standardjs should be moaning about this

**Review by samplackett — COMMENTED** [review]
One small tweak suggested.

Also, as with other ones past few days, I think our Jenkins pipeline is never going to succeed until we iterate the version number.

File: app/publishing/save-request.js [hunk 2]
```diff
@@ -3,17 +3,19 @@ const db = require('../data')
...
   try {
     const timestamp = new Date()
     const statement = await saveStatement(request, timestamp, transaction)
     const delivery = await saveDelivery(statement.statementId, method, reference, timestamp, transaction)
-    if (reason) {
-      console.log(`Unable to deliver statement ${statement.filename} to "${statement.email}": ${reason}`)
-      await sendCrmMessage(statement.email, statement.frn, reason)
-      await createFailure(delivery.deliveryId, reason, transaction)
+    
+    if (errorObject.reason) {
```
Comment (jbarnardeviden) [inline app/publishing/save-request.js:15]: I'm not sure this is right; reason is set to undefined unless the error is triggered by the email validation rule i.e. we might not be handling a rejection from the API itself here.

File: app/publishing/publish-statement.js [hunk 3]
```diff
@@ -24,13 +25,20 @@ const publishStatement = async (request) => {
   try {
     validateEmail(request.email)
     const personalisation = getPersonalisation(request.scheme.name, request.scheme.shortName, request.scheme.year, request.scheme.frequency, request.businessName, request.paymentPeriod)
+    //TODO Add check if email errored out
```
Comment (jbarnardeviden) [inline app/publishing/publish-statement.js:28]: We catching errors here but will also need to check the Notify API for status updates periodically.

File: app/publishing/publish-by-email.js [hunk 4]
```diff
@@ -1,19 +1,70 @@
...
+ * @returns Thunk
+ */
+function setupEmailThunk(template, email, file, personalisation, notifyClient){
   const latestDownloadDate = moment(new Date()).add(config.retentionPeriodInWeeks, 'weeks').format('LL')
-  const notifyAttachment = await notifyClient.prepareUpload(file, { confirmEmailBeforeDownload: true, retentionPeriod: `${config.retentionPeriodInWeeks} weeks` })
-  return notifyClient.sendEmail(template, email, {
+  return promisify(thunkify(notifyClient.sendEmail, template, email, {
     personalisation: {
-      link_to_file: notifyAttachment,
+      link_to_file: notifyClient.prepareUpload(
```
Comment (jbarnardeviden) [inline app/publishing/publish-by-email.js:43]: flawed: this promise needs to be resolved first...

**Pull request merged**